{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to Databases and Structured Query Language (SQL)\n",
    "\n",
    "As Data Scientists, you will frequently want to store data in an organized, structured manner that allows you to do complex queries.  Because you are good Data Scientists, [**you do not use Excel!!!**](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-5-80)\n",
    "\n",
    "In this course, we will only discuss **Relational Databases**, because those are the most common in bioinformatics.  (There are other kinds!!).  So when I say \"database\" I mean \"relational database\".\n",
    "\n",
    "Databases are used to store information in a manner that, when used properly, is:\n",
    "  a) highly structured\n",
    "  b) constrained (i.e. detects errors)\n",
    "  c) transactional (i.e. can undo a command if it discovers a problem)\n",
    "  d) indexed (for speed of search)\n",
    "  e) searchable\n",
    "  \n",
    "The core concept of a database is a **Table**.  Tables contain one particular \"kind\" of information (e.g. a Table could represent a Student, a University, a Book, or a Clinical Record.  \n",
    "\n",
    "Tables contain **Rows** and **Columns** where, generally, every column represents a \"feature\" of that information (e.g. a Student table might have **[\"name\", \"gender\", \"studentID\", \"age\"]** as its columns/features).  Every row represents an \"individual\", and their values for each feature (e.g. a Row in a Student table might have **[\"Mark Wilkinson\", \"M\", \"163483\", \"35\"]** as its values.\n",
    "\n",
    "A Database may have many Tables that represent various kinds of related information.  For example, a library database might have a Books table, a Publishers table, and a Locations table.  A Book has a Publisher, and a Location, so the tables need to be connected to one another.  This is achieved using **keys**.  Generally, every row (individual) in a table has a unique identifier (generally a number), and this is called its **key**.  Because it is unique, it is possible to refer unambiguously to that individual record.\n",
    "\n",
    "I think the easiest way to learn about databases and SQL is to start building one!  We will use the MySQL Docker Container that we created in the previous lesson.  We are going to create a Germplasm database (seed stocks).  It will contain information about the seed (its amount, its harvest date, its location), the germplasm (its species, the allele it carries), and about the genetics related to that allele (the gene_id, the gene name, the protein name, and a link to the GenBank record)\n",
    "\n",
    "(if that container isn't running, please **docker start course-mysql** now!)\n",
    "\n",
    "**Note:  This Jupyter Notebook is running the Python kernel.  This allows us to use some nice tools in Python (the sql extension and SqlMagic) that provide access to the mysql database server from inside of the Notebook.  You don't need to know any Python to do this.  Note also that you can do exactly the same commands in your Terminal window.**\n",
    "\n",
    "To connect to the MySQL Docker Container from your terminal window, type:\n",
    "\n",
    "     mysql -h 127.0.0.1 -P 3306 --protocol=tcp -u root -p\n",
    " \n",
    "(then enter your password 'root' to access the database)\n",
    " \n",
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "# SQL\n",
    "\n",
    "Structured Query Language is a way to interact with a database server.  It is used to create, delete, edit, fill, and query tables and their contents.  \n",
    "\n",
    "First, we will learn the SQL commands that allow us to explore the database server, and create new databases and tables..  Later, we will use SQL to put information into those tables.  Finally, we will use SQL to query those tables.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Python SQL Extension\n",
    "\n",
    "The commands below are used to connect to the MySQL server in our Docker Container.  You need to execute them ONCE.  In every subsequent Juputer code window, you will have access to the database.\n",
    "\n",
    "all SQL commands are preceded by \n",
    "\n",
    "     %sql \n",
    "     \n",
    "(**only in the Python extension!  Not in your terminal window!**)\n",
    "\n",
    "all SQL commands end with a \";\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n",
    "#%sql mysql+pymysql://anonymous@ensembldb.ensembl.org/homo_sapiens_core_92_38\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## show databases\n",
    "\n",
    "**show databases** is the command to see what databases exist in the server.  The ones you see now are the default databases that MySQL uses to organize itself.  _**DO NOT TOUCH THESE DATABASES**_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "4 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>information_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>mysql</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>performance_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>sys</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('information_schema',), ('mysql',), ('performance_schema',), ('sys',)]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql show databases;\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## create database\n",
    "\n",
    "The command to create a database is **create database** (surprise!  ;-) )\n",
    "\n",
    "We will create a database called \"germplasm\"\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>information_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>mysql</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>performance_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>sys</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('information_schema',),\n",
       " ('germplasm',),\n",
       " ('mysql',),\n",
       " ('performance_schema',),\n",
       " ('sys',)]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql create database germplasm;\n",
    "%sql show databases\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## use database_name\n",
    "\n",
    "the **use** command tells the server which database you want to interact with.  Here we will use the database we just created"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql use germplasm\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## show tables\n",
    "\n",
    "The show tables command shows what tables the database contains (right now, none!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Tables_in_germplasm</th>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "% sql show tables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Planning your data structure\n",
    "\n",
    "This is the hard part.  What does our data \"look like\" in a well-structured, relational format?\n",
    "\n",
    "Starting simply:\n",
    "\n",
    "<center>stock table</center>\n",
    "\n",
    "  amount  |  date  |  location  \n",
    " --- | --- | --- \n",
    "  5  | 10/5/2013 |  Room 2234  \n",
    "  9.8  | 12/1/2015 |  Room 998  \n",
    "\n",
    "\n",
    "-----------------------------\n",
    "\n",
    "\n",
    "<center>germplasm table</center>\n",
    "\n",
    "  taxonid  |  allele\n",
    " --- | --- \n",
    "  4150  | def-1\n",
    "  3701  | ap3\n",
    "  \n",
    "--------------------------------\n",
    "\n",
    "<center>gene table</center>\n",
    "\n",
    "  gene  |  gene_name  |  embl\n",
    " --- | ---  | --- \n",
    "  DEF  | Deficiens  | https://www.ebi.ac.uk/ena/data/view/AB516402\n",
    "  AP3  | Apetala3   |   https://www.ebi.ac.uk/ena/data/view/AF056541\n",
    "  \n",
    "  \n",
    "  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## add indexes\n",
    "\n",
    "It is usually a good idea to have an index column on every table, so let's add that first:\n",
    "\n",
    "\n",
    "<center>stock table</center>\n",
    "\n",
    "id  |  amount  |  date  |  location  \n",
    "---  | --- | --- | --- \n",
    "1   |  5  | 10/5/2013 |  Room 2234  \n",
    "2   |   9.8  | 12/1/2015 |  Room 998  \n",
    "\n",
    "\n",
    "-----------------------------\n",
    "\n",
    "\n",
    "<center>germplasm table</center>\n",
    "\n",
    "id  |  taxonid  |  allele\n",
    "--- |  --- | --- \n",
    "1  |  4150  | def-1\n",
    "2  |   3701  | ap3\n",
    "  \n",
    "--------------------------------\n",
    "\n",
    "<center>gene table</center>\n",
    "\n",
    "id  |  gene  |  gene_name  |  embl\n",
    "--- | --- | ---  | --- \n",
    "1  |  DEF  | Deficiens  | https://www.ebi.ac.uk/ena/data/view/AB516402\n",
    "2  |  AP3  | Apetala3   |   https://www.ebi.ac.uk/ena/data/view/AF056541\n",
    "  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  find linkages\n",
    "\n",
    "* Every germplasm has a stock record.  This is a 1:1 relationship.\n",
    "* Every germplasm represents a specific gene.  This is a 1:1 relationship\n",
    "\n",
    "So every germplasm must point to the index of a stock, and also to the index of a gene\n",
    "\n",
    "Adding that into our tables we have:\n",
    "\n",
    "\n",
    "\n",
    "<center>stock table</center>\n",
    "\n",
    "id  |  amount  |  date  |  location  \n",
    "---  | --- | --- | --- \n",
    "1   |  5  | 10/5/2013 |  Room 2234  \n",
    "2   |   9.8  | 12/1/2015 |  Room 998  \n",
    "\n",
    "\n",
    "-----------------------------\n",
    "\n",
    "\n",
    "<center>germplasm table</center>\n",
    "\n",
    "id  |  taxonid  |  allele  |  stock_id  |  genetics_id\n",
    "--- |  --- | ---  | --- | ---\n",
    "1  |  4150  | def-1  | 2   |  1\n",
    "2  |   3701  | ap3   | 1   |  2\n",
    "  \n",
    "--------------------------------\n",
    "\n",
    "<center>gene table</center>\n",
    "\n",
    "id  |  gene  |  gene_name  |  embl\n",
    "--- | --- | ---  | --- \n",
    "1  |  DEF  | Deficiens  | https://www.ebi.ac.uk/ena/data/view/AB516402\n",
    "2  |  AP3  | Apetala3   |   https://www.ebi.ac.uk/ena/data/view/AF056541\n",
    "  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## data types in MySQL\n",
    "\n",
    "I will not discuss [all MySQL Datatypes](https://dev.mysql.com/doc/refman/5.7/en/data-types.html), but we will look at only the ones we need.  We need:\n",
    "\n",
    "* Integers (type INTEGER)\n",
    "* Floating point (type FLOAT)\n",
    "* Date  (type DATE [in yyyy-mm-dd format](https://dev.mysql.com/doc/refman/5.7/en/datetime.html) )\n",
    "* Characters (small, variable-length --> type [VARCHAR(x)](https://dev.mysql.com/doc/refman/5.7/en/char.html) )\n",
    "\n",
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "## create table \n",
    "\n",
    "tables are created using the **create table** command (surprise!)\n",
    "\n",
    "The [syntax of create table](https://dev.mysql.com/doc/refman/5.7/en/create-table.html) can be quite complicated, but we are only going to do the most simple examples.\n",
    "\n",
    "    create table table_name (column_name column_definition)\n",
    "    \n",
    "column definitions include the data-type, and other options like if it is allowed to be null(blank), or if it should be treated as an \"index\" column.\n",
    "\n",
    "Examples are easier to understand than words... so here are our table definitions:\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "4 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td>auto_increment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>amount</td>\n",
       "        <td>float</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>date</td>\n",
       "        <td>date</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>location</td>\n",
       "        <td>varchar(20)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('id', 'int(11)', 'NO', 'PRI', None, 'auto_increment'),\n",
       " ('amount', 'float', 'NO', '', None, ''),\n",
       " ('date', 'date', 'NO', '', None, ''),\n",
       " ('location', 'varchar(20)', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#%sql drop table stock\n",
    "%sql create table stock(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, amount FLOAT NOT NULL, date DATE NOT NULL, location VARCHAR(20) NOT NULL);\n",
    "%sql describe stock\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td>auto_increment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>taxonid</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>allele</td>\n",
       "        <td>varchar(10)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>stock_id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>gene_id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('id', 'int(11)', 'NO', 'PRI', None, 'auto_increment'),\n",
       " ('taxonid', 'int(11)', 'NO', '', None, ''),\n",
       " ('allele', 'varchar(10)', 'NO', '', None, ''),\n",
       " ('stock_id', 'int(11)', 'NO', '', None, ''),\n",
       " ('gene_id', 'int(11)', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#%sql drop table germplasm\n",
    "%sql create table germplasm(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, taxonid INTEGER NOT NULL, allele VARCHAR(10) NOT NULL, stock_id INTEGER NOT NULL, gene_id INTEGER NOT NULL);\n",
    "%sql describe germplasm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "4 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td>auto_increment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>gene</td>\n",
       "        <td>varchar(10)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>gene_name</td>\n",
       "        <td>varchar(30)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>embl</td>\n",
       "        <td>varchar(70)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('id', 'int(11)', 'NO', 'PRI', None, 'auto_increment'),\n",
       " ('gene', 'varchar(10)', 'NO', '', None, ''),\n",
       " ('gene_name', 'varchar(30)', 'NO', '', None, ''),\n",
       " ('embl', 'varchar(70)', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#%sql drop table gene\n",
    "%sql create table gene(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, gene VARCHAR(10) NOT NULL, gene_name VARCHAR(30) NOT NULL, embl VARCHAR(70) NOT NULL);\n",
    "%sql describe gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## loading data\n",
    "\n",
    "There are many ways to import data into MySQL.  If you have data in another (identical) MySQL database, you can \"dump\" the data, and then import it directly.  If you have tab or comma-delimited (tsv, csv) you can **sometimes** import it directly from these formats.  You can also enter data using SQL itself.  This is usually the safest way, when you have to keep multiple tables synchronized (as we do, since the germplasm table is \"linked to\" the other two tables)\n",
    "\n",
    "## insert into\n",
    "\n",
    "The command to load data is:\n",
    "\n",
    "    insert into table_name (field1, field2, field3) values (value1, value2, value3)\n",
    "    \n",
    "Now... what data do we need to add, in what order?\n",
    "\n",
    "The germplasm table needs the ID number from both the gene table and the stock table, so we cannot enter the germplasm information first.  We must therefore enter the gene and stock data first.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: Line magic function `%select` not found.\n"
     ]
    }
   ],
   "source": [
    "# NOTE - we DO NOT put data into the \"id\" column!  This column is auto_increment, so it \"magically\" creates its own value\n",
    "%sql insert into gene (gene, gene_name, embl) values ('DEF', \"Deficiens\", 'https://www.ebi.ac.uk/ena/data/view/AB516402');\n",
    "%sql insert into gene (gene, gene_name, embl) values ('AP3', \"Apetala3\", 'https://www.ebi.ac.uk/ena/data/view/AF056541');\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>last_insert_id()</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>2</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(2,)]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql select last_insert_id();  # just to show you that this function exists!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql insert into stock(amount, date, location) values (5, '2013-05-10', 'Room 2234');\n",
    "%sql insert into stock(amount, date, location) values (9.8, '2015-1-12', 'Room 998');\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Almost ready!  \n",
    "\n",
    "We now need to know the index numbers from the stock and gene databases that correspond to the data for the germplasm table.  For this, we need to learn another function:  **select**\n",
    "\n",
    "## Select statements\n",
    "\n",
    "**Select** is the command used to query the database.  We will look in more detail later, but now all you need to know is that the most basic structure is:\n",
    "\n",
    "     select * from table_name\n",
    "     \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>id</th>\n",
       "        <th>amount</th>\n",
       "        <th>date</th>\n",
       "        <th>location</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>1</td>\n",
       "        <td>5.0</td>\n",
       "        <td>2013-05-10</td>\n",
       "        <td>Room 2234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>2</td>\n",
       "        <td>9.8</td>\n",
       "        <td>2015-01-12</td>\n",
       "        <td>Room 998</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(1, 5.0, datetime.date(2013, 5, 10), 'Room 2234'),\n",
       " (2, 9.8, datetime.date(2015, 1, 12), 'Room 998')]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql select * from stock;  # notice that the id number was automatically generated\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>id</th>\n",
       "        <th>gene</th>\n",
       "        <th>gene_name</th>\n",
       "        <th>embl</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>1</td>\n",
       "        <td>DEF</td>\n",
       "        <td>Deficiens</td>\n",
       "        <td>https://www.ebi.ac.uk/ena/data/view/AB516402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>2</td>\n",
       "        <td>AP3</td>\n",
       "        <td>Apetala3</td>\n",
       "        <td>https://www.ebi.ac.uk/ena/data/view/AF056541</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(1, 'DEF', 'Deficiens', 'https://www.ebi.ac.uk/ena/data/view/AB516402'),\n",
       " (2, 'AP3', 'Apetala3', 'https://www.ebi.ac.uk/ena/data/view/AF056541')]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql select * from gene;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<pre>\n",
    "\n",
    "\n",
    "</pre>\n",
    "\n",
    "Just a reminder, our germplasm data is:\n",
    "\n",
    "id  | \ttaxonid  |  allele  | \tstock_id  |\tgene_id\n",
    "--- | --- | --- | --- | --- |\n",
    "1 \t | 4150 \t|def-1 \t\n",
    "2 \t| 3701 |\tap3 \t"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
